### PR TITLE
Add Support for Symfony 5.4

### PR DESCRIPTION
**Pull request by AdamKasp**
| Q               | A
| --------------- | -----
| Branch?         |  master
| Bug fix?        | no
| New feature?    |yes
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

### CODE REVIEW

File: config/packages/security.yaml [hunk 1]
```diff
@@ -1,5 +1,4 @@
 security:
-    always_authenticate_before_granting: true
```
Comment (AdamKasp) [inline config/packages/security.yaml:2]: this flag is deprecated in Symfony 5.4 and will be removed in Symfony 6.0. what is more, it breaks application after bumping the version to Symfony 5.4
references:
- https://symfony.com/doc/5.4/reference/configuration/security.html#always-authenticate-before-granting
- https://github.com/symfony/symfony/issues/44337

Comment (lchrusciel) [reply config/packages/security.yaml:2]: Can we test this behaviour? https://github.com/Sylius/Sylius/pull/10289#issuecomment-480249379

Still, this is something that should be placed in UPGRADE file for others to make an upgrade, as this change will not take any effect over 3rd party users

File: src/Sylius/Bundle/ApiBundle/Controller/UploadAvatarImageAction.php [hunk 2]
```diff
@@ -57,7 +57,7 @@ public function __invoke(Request $request): ImageInterface
         $image->setFile($file);
 
         /** @var string $ownerIri */
-        $ownerIri = $request->get('owner');
+        $ownerIri = $request->request->get('owner');
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Controller/UploadAvatarImageAction.php:60]: This could be considered as BC-Break, but I'm fine with that. Can we add a note about it?

File: src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php [hunk 3]
```diff
@@ -38,7 +38,10 @@ public function __construct(
     public function __invoke(Request $request): JsonResponse
     {
         /** @var PaymentInterface|null $payment */
-        $payment = $this->paymentRepository->findOneByOrderToken($request->get('paymentId'), $request->get('id'));
+        $payment = $this->paymentRepository->findOneByOrderToken(
+            $request->attributes->get('paymentId'),
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php:42]: This could be considered as BC-Break, but I'm fine with that. Can we add a note about it?

File: src/Sylius/Bundle/CoreBundle/Controller/ProductTaxonController.php [hunk 4]
```diff
@@ -103,4 +107,27 @@ private function updatePositions(string $position, int $id): void
...
+     * based on Symfony\Component\HttpFoundation\Request::get
+     */
+    private function getParameterFromRequest(Request $request, string $key)
+    {
+        if ($request !== $result = $request->attributes->get($key, $request)) {
+            return $result;
+        }
+
+        if ($request->query->has($key)) {
+            return $request->query->all()[$key];
```
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/Controller/ProductTaxonController.php:124]: Why not this way?
```suggestion
            return $request->query->get($key);
```

File: src/Sylius/Bundle/CoreBundle/Controller/ProductTaxonController.php [hunk 5]
```diff
@@ -103,4 +107,27 @@ private function updatePositions(string $position, int $id): void
...
+
+    /**
+     * @return mixed
+     *
+     * @deprecated This function will be removed in Sylius 2.0, since Symfony 5.4, use explicit input sources instead
+     * based on Symfony\Component\HttpFoundation\Request::get
+     */
+    private function getParameterFromRequest(Request $request, string $key)
+    {
+        if ($request !== $result = $request->attributes->get($key, $request)) {
```
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/Controller/ProductTaxonController.php:119]: Why do we have to do it this way? It is not obvious what is the expected result. Why can't we use `has` method?

File: src/Sylius/Bundle/CoreBundle/Controller/ProductTaxonController.php [hunk 6]
```diff
@@ -103,4 +107,27 @@ private function updatePositions(string $position, int $id): void
...
+            return $result;
+        }
+
+        if ($request->query->has($key)) {
+            return $request->query->all()[$key];
+        }
+
+        if ($request->request->has($key)) {
+            return $request->request->all()[$key];
```
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/Controller/ProductTaxonController.php:128]:
```suggestion
            return $request->request->get($key);
```

File: src/Sylius/Bundle/CoreBundle/Controller/ProductVariantController.php [hunk 7]
```diff
@@ -54,4 +55,27 @@ public function updatePositionsAction(Request $request): Response
...
         return new JsonResponse();
     }
+
+    /**
+     * @return mixed
+     *
+     * @deprecated This function will be removed in Sylius 2.0, since Symfony 5.4, use explicit input sources instead
+     * based on Symfony\Component\HttpFoundation\Request::get
+     */
+    private function getParameterFromRequest(Request $request, string $key)
```
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/Controller/ProductVariantController.php:65]: I'm wondering if we shouldn't port it down to ResourceBundle

File: src/Sylius/Bundle/ApiBundle/Controller/UploadAvatarImageAction.php (hunk 2 above)
Comment (AdamKasp) [reply src/Sylius/Bundle/ApiBundle/Controller/UploadAvatarImageAction.php:60]: we have `@experimental` on the entire apiBundle so IMO it isn't BC-break, but ofc I will add a note to apiUpgrade file

File: src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php (hunk 3 above)
Comment (AdamKasp) [reply src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php:42]: we have `@experimental` on the entire apiBundle so IMO it isn't BC-break, but ofc I will add a note to apiUpgrade file

File: src/Sylius/Bundle/CoreBundle/Controller/ProductTaxonController.php (hunk 5 above)
Comment (AdamKasp) [reply src/Sylius/Bundle/CoreBundle/Controller/ProductTaxonController.php:119]: it is a workaround over usage of internal function from `Symfony\Component\HttpFoundation\Request` and it is an almost copy-paste of function `get` from this service. I intentionally don't refactored it, because this is temporary and will be removed in Sylius 2.0

File: src/Sylius/Bundle/CoreBundle/Controller/ProductTaxonController.php (hunk 6 above)
Comment (AdamKasp) [reply src/Sylius/Bundle/CoreBundle/Controller/ProductTaxonController.php:128]: https://github.com/Sylius/Sylius/pull/13339#discussion_r759172172

File: src/Sylius/Bundle/CoreBundle/Controller/ProductVariantController.php (hunk 7 above)
Comment (AdamKasp) [reply src/Sylius/Bundle/CoreBundle/Controller/ProductVariantController.php:65]: it is temporary logic, I am not sure if we need to make bigger refactor of it.

File: tests/Controller/AdminProductAjaxTest.php [hunk 8]
```diff
@@ -65,7 +65,7 @@ private function authenticateAdminUser(): void
         $session = self::$container->get('session');
         $firewallName = 'admin';
         $firewallContext = 'admin';
-        $token = new UsernamePasswordToken($user, null, $firewallName, $user->getRoles());
+        $token = new UsernamePasswordToken($user, $firewallName, $user->getRoles());
```
Comment (mbabker) [inline tests/Controller/AdminProductAjaxTest.php:68]: You still need the credentials parameter for Symfony 4.4.

```php
// Symfony 5.4 deprecates the `$credentials` argument of the token class
if (3 === (new \ReflectionClass(UsernamePasswordToken::class))->getConstructor()->getNumberOfParameters()) {
    $token = new UsernamePasswordToken($user, $firewallName, $user->getRoles());
} else {
    $token = new UsernamePasswordToken($user, null, $firewallName, $user->getRoles());
}
```

File: src/Sylius/Bundle/CoreBundle/Controller/ProductVariantController.php (hunk 7 above)
Comment (lchrusciel) [reply src/Sylius/Bundle/CoreBundle/Controller/ProductVariantController.php:65]: this "temporary" logic will be with us for a long time

**Review by lchrusciel — APPROVED** [review]

File: src/Sylius/Behat/Service/SecurityService.php [hunk 9]
```diff
@@ -43,7 +43,15 @@ public function __construct(SessionInterface $session, CookieSetterInterface $co
 
     public function logIn(UserInterface $user): void
     {
-        $token = new UsernamePasswordToken($user, $user->getPassword(), $this->firewallContextName, $user->getRoles());
+        /**
+         * @deprecated parameter credential was deprecated in Symfony 5.4, so in Sylius 1.11 too, in Sylius 2.0 providing 4 arguments will be prohibited.
+         */
```
Comment (GSadee) [inline src/Sylius/Behat/Service/SecurityService.php:48]:
```suggestion
        /** @deprecated parameter credential was deprecated in Symfony 5.4, so in Sylius 1.11 too, in Sylius 2.0 providing 4 arguments will be prohibited. */
```

File: src/Sylius/Bundle/CoreBundle/Security/UserImpersonator.php [hunk 10]
```diff
@@ -40,12 +40,24 @@ public function __construct(SessionInterface $session, string $firewallContextNa
...
+        if (3 === (new \ReflectionClass(UsernamePasswordToken::class))->getConstructor()->getNumberOfParameters()) {
+            $token = new UsernamePasswordToken(
+                $user,
+                $this->firewallContextName,
+                array_map(/** @param object|string $role */ static function ($role): string { return (string) $role; }, $user->getRoles())
+            );
+        } else {
+            $token = new UsernamePasswordToken(
+                $user,
+                null,
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Security/UserImpersonator.php:55]: shouldn't it stay like before?
```suggestion
                $user->getPassword(),
```

File: src/Sylius/Bundle/ProductBundle/Controller/ProductSlugController.php [hunk 11]
```diff
@@ -13,17 +13,38 @@
...
+    private ?SlugGeneratorInterface $slugGenerator;
+
+    public function __construct(?SlugGeneratorInterface $slugGenerator = null)
+    {
+        $this->slugGenerator = $slugGenerator;
+
+        if ($this->slugGenerator === null) {
+            @trigger_error(sprintf('Not passing a $slugGenerator to %s constructor is deprecated since Sylius 1.11 and will be prohibited in Sylius 2.0.', self::class), \E_USER_DEPRECATED);
+        }
+    }
```
Comment (GSadee) [inline src/Sylius/Bundle/ProductBundle/Controller/ProductSlugController.php:33]: Why did you add this service in this PR?

Comment (AdamKasp) [reply src/Sylius/Bundle/ProductBundle/Controller/ProductSlugController.php:33]: because
Screenshot 2021-12-02 at 10 39 50
 in Symfony 5.4 function `get` is deprecated

File: src/Sylius/Bundle/CoreBundle/Controller/ProductVariantController.php (hunk 7 above)
Comment (Zales0123) [reply src/Sylius/Bundle/CoreBundle/Controller/ProductVariantController.php:65]: 👍 for moving it to `ResourceBundle`. However, to not stop this PR for too long, I suggest we merge it and proceed with `ResourceBundle` changes parallelly 🖖

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thank you, Adam! :tada:
